### PR TITLE
feat: annotations REST API

### PR DESCRIPTION
**Pull request by dpgaspar**
### SUMMARY
New REST API for `Annotations` and `Annotation Layers`, implements basic CRUD and bulk select

Integrity:
- Refuses to delete a layer that has associated annotations (this was possible before)
- checks types and lengths (basic marshmallow)
- Validates annotations JSON
- Validates end datetime is after start datetime is defined
- validates layer name uniqueness
- validates annotation short description uniqueness on the same layer

Screenshot 2020-10-20 at 16 00 21

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by willbarrett — COMMENTED** [review]
Overall this looks really good, a couple minor comments.

File: superset/annotation_layers/annotations/dao.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    @staticmethod
+    def validate_uniqueness(layer_id: int, short_descr: str) -> bool:
+        query = db.session.query(Annotation).filter(
+            Annotation.short_descr == short_descr, Annotation.layer_id == layer_id
+        )
+        return not db.session.query(query.exists()).scalar()
+
+    @staticmethod
+    def validate_update_uniqueness(
```
[the remark this answers is not in the document]
Comment (dpgaspar) [reply superset/annotation_layers/annotations/dao.py:55]: I like that

File: superset/annotation_layers/annotations/dao.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        annotation_id, layer_id: int, short_descr: str
```
Comment (villebro) [inline superset/annotation_layers/annotations/dao.py:56]: nit:
```suggestion
        annotation_id: int, layer_id: int, short_descr: str
```

File: tests/core_tests.py [hunk 3]
```diff
@@ -199,11 +199,16 @@ def test_annotation_json_endpoint(self):
         self.assertIn("id", resp_annotations["result"][0])
         self.assertIn("name", resp_annotations["result"][0])
 
-        layer = self.get_resp(
+        response = self.get_resp(
             f"/superset/annotation_json/{layer.id}?form_data="
             + quote(json.dumps({"time_range": "100 years ago : now"}))
         )
-        self.assertIn("my_annotation", layer)
+        self.assertIn("my_annotation", response)
```
Comment (villebro) [inline tests/core_tests.py:206]: nit, while at it, perhaps do
```suggestion
        assert "my_annotation" in response
```

**Review by villebro — APPROVED** [review]
Looking super solid, great work! Left a few trivial comments.

File: superset/annotation_layers/annotations/commands/create.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    def validate(self) -> None:
+        exceptions: List[ValidationError] = list()
+        layer_id: Optional[int] = self._properties.get("layer")
+        start_dttm: Optional[datetime] = self._properties.get("start_dttm")
+        end_dttm: Optional[datetime] = self._properties.get("end_dttm")
+        short_descr = self._properties.get("short_descr", "")
+
+        # Validate/populate model exists
+        if not layer_id and not isinstance(layer_id, int):
+            raise AnnotationLayerNotFoundError
```
Comment (dpgaspar) [inline superset/annotation_layers/annotations/commands/create.py:63]: missing instantiation

**Pull request merged**